plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,425 @@
+				guint write_cnt,
+				guint8 *read_buf,
+				guint *read_cnt,
+				GError **error)
+{
+	const gsize bufsz = 0x40; /* 8 bytes read packet */
+	guint8 *buf = g_malloc0(bufsz);
+
+	if (!mediatek_scaler_device_ddc_write(self, write_buf, write_cnt, error))
+		return FALSE;
+
+	g_usleep(FU_MEDIATEK_SCALER_DDC_MSG_DELAY * 1000);
+	if (!fu_udev_device_pread(FU_UDEV_DEVICE(self->i2c_dev), 0x0, buf, bufsz, error)) {
+		return FALSE;
+	}
+
+	if (buf[0] == DDC_DEST_ADDR && buf[1] > DDC_DATA_LEN_DFT) {
+		gsize recvsz = buf[1] - DDC_DATA_LEN_DFT;
+		guint8 checksum = 0;
+		checksum ^= 0x50; /* slave source address */
+		for (gsize i = 0; i < recvsz + 2; i++)
+			checksum ^= buf[i];
+
+		if (checksum != buf[recvsz + 2])
+			return FALSE;
+
+		memcpy(read_buf, buf, recvsz);
+		*read_cnt = (guint)recvsz;
+		return TRUE;
+	}
+
+	g_set_error(error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_READ,
+		    "DDC READ over %s results invalid: [0] expected '0x%02x', got '0x%02x'; [1] "
+		    "expected '0x%02x' gr '0x%02x'",
+		    self->i2c_dev_name,
+		    (guint)DDC_DEST_ADDR,
+		    buf[0],
+		    buf[1],
+		    (guint)DDC_DATA_LEN_DFT);
+	return FALSE;
+}
+
+static gboolean
+mediatek_scaler_display_is_connected(FuMediatekScalerDevice *self, GError **error)
+{
+	guint8 value1 = g_random_int_range(1, 255);
+	guint8 value2 = g_random_int_range(1, 255);
+	guint8 sum = value1 + value2;
+	guint8 buf[] = {0xCC, 0xFE, value1, value2};
+	g_autofree guint8 *recv = g_malloc0(10);
+	g_autoptr(GError) error_local = NULL;
+
+	if (mediatek_scaler_device_ddc_read(self, buf, sizeof(buf), recv, NULL, &error_local)) {
+		if (sum == recv[3])
+			return TRUE;
+	}
+
+	g_set_error(error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_NOT_SUPPORTED,
+		    "Ignored display '%s': %s",
+		    self->dp_aux_dev_name,
+		    error_local->message);
+	return FALSE;
+}
+
+static gboolean
+mediatek_scaler_device_probe_firmware_version(FuDevice *device, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autofree guint8 *recv = g_malloc0(10);
+	guint8 buf[] = {0xCC, 0xFF, 0x01};
+	guint recvsz = 0;
+
+	if (!mediatek_scaler_device_ddc_read(self, buf, sizeof(buf), recv, &recvsz, error))
+		return FALSE;
+
+	if (recvsz >= 4) {
+		g_autofree guint8 *ver = g_new0(guint8, 4);
+		g_autofree guint32 *ver32 = (guint32 *)ver;
+		memcpy(ver + 0, &recv[4], sizeof(guint8));
+		memcpy(ver + 1, &recv[5], sizeof(guint8));
+		memcpy(ver + 2, &recv[2], sizeof(guint8));
+		memcpy(ver + 3, &recv[3], sizeof(guint8));
+
+		fu_device_set_version_from_uint32(FU_DEVICE(self), *ver32);
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static gboolean
+mediatek_scaler_device_probe_hardware_version(FuDevice *device, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autofree guint8 *recv = g_malloc0(10);
+	guint8 buf[] = {0xCC, 0xFF, 0x00};
+	guint recvsz = 0;
+
+	if (!mediatek_scaler_device_ddc_read(self, buf, sizeof(buf), recv, &recvsz, error))
+		return FALSE;
```

Comment:
I think passing a GByteArray as the buffer to read and a GByteArray as the buffer to read would make this a lot safer and be easier to read.